### PR TITLE
added inline ``sorted by`` expression support with InputSource and as…

### LINKED ISSUE #8366: Add an in-line sort by expression
### Suggestion

Add an in-lined `sort x in (ascending|descending) orer by y` expression that functions the same way as its counterpart effect

### Why?

Quite often I find myself making a copy of a variable simply to sort it and reorder the variable I just made a copy of.
Adding an expression that allows me to cut out that extra step just helps save time in that aspect.

I never understood why the `ExprSorted` never received an updated pattern to follow the behavior of `EffSort`

### Other

Example skript for how we currently are forced to handle this
```
function example() :: objects:
    copy {-myStoredObjects::*} into {_myChangedObjects::*}
    sort  {_myChangedObjects::*} in descending order by (name of input)
    return {_myChangedObjects::*}
```

where this could have instead been
```
function example() :: objects:
    return {-myStoredObjects::*} sorted in descending order by (name of input)
```

### Agreement

- [x] I have read the guidelines above and affirm I am following them with this suggestion.

### PR DESCRIPTION
**Pull request by ethanrjs**
### Problem
`ExprSortedList` only supports basic natural-order sorting via `sorted %objects%`. To sort with a custom comparator or in descending order, you have to use `EffSort`, which mutates in-place, meaning you're forced to copy the list first:
```sk
copy {-myStoredObjects::*} into {_temp::*}
sort {_temp::*} in descending order by (name of input)
return {_temp::*}
```

### Solution
Added `sorted ... by` expression patterns to `ExprSortedList` with `InputSource` support, matching `EffSort`'s behavior. The existing `sorted %objects%` pattern is unchanged.
```sk
return {-myStoredObjects::*} sorted in descending order by (name of input)
```

### Testing Completed
Added `ExprSortedList.sk` test file covering basic sorting, ascending/descending with sort-by, and edge cases.
image

Also ran tests on-server.
image

### Supporting Information
First contribution. Open to feedback :).

---
**Completes:** #8366
**Related:** #6158
**AI assistance:** Opus 4.6 Max reviewed my code + wrote tests in `ExprSortedList.sk`

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 1]
```diff
@@ -15,38 +13,64 @@
...
+public class ExprSortedList extends SimpleExpression<Object> implements KeyedIterableExpression<Object>, InputSource {
+
+	private record MappedValue(Object original, Object mapped) { }
+	private record KeyedMappedValue(KeyedValue<?> keyed, Object mapped) { }
 
 	static {
-		Skript.registerExpression(ExprSortedList.class, Object.class, ExpressionType.PROPERTY, "sorted %objects%");
+		Skript.registerExpression(ExprSortedList.class, Object.class, ExpressionType.PROPERTY,
+			"sorted %objects%",
+			"%objects% sorted [in (:descending|ascending) order] [(by|based on) <.+>]"
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: please see ExprFilter for adding required () or [] around the <> portion of the pattern. We require these in nested pattern expressions to help the parser avoid self-immolation

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 2]
```diff
@@ -57,15 +81,64 @@ public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean is
...
-			return list.stream(event)
-					.sorted(ExprSortedList::compare)
+		int sortingMultiplier = descendingOrder ? -1 : 1;
+		if (mappingExpr == null) {
+			try {
+				return list.stream(event)
+					.sorted((o1, o2) -> compare(o1, o2) * sortingMultiplier)
 					.toArray();
+			} catch (IllegalArgumentException | ClassCastException e) {
+				return (Object[]) Array.newInstance(getReturnType(), 0);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: should probably be throwing a runtime error here

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 3]
```diff
@@ -57,15 +81,64 @@ public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean is
...
+
+		List<MappedValue> mappedValues = new ArrayList<>();
+		if (keyed) {
+			for (Iterator<? extends KeyedValue<?>> it = ((KeyedIterableExpression<?>) list).keyedIterator(event); it.hasNext(); ) {
+				KeyedValue<?> keyedValue = it.next();
+				currentIndex = keyedValue.key();
+				currentValue = keyedValue.value();
+				Object mappedValue = mappingExpr.getSingle(event);
+				if (mappedValue == null)
+					return (Object[]) Array.newInstance(getReturnType(), 0);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: error here too

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 4]
```diff
@@ -57,15 +81,64 @@ public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean is
...
+			if (!mappingExpr.isSingle()) {
+				Skript.error("The mapping expression in the sort expression must only return a single value for a single input.");
+				return false;
+			}
+		}
 		return LiteralUtils.canInitSafely(list);
 	}
 
 	@Override
 	protected Object @Nullable [] get(Event event) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: the logic in here should be moved to iterator() and get should just delegate to either iterator or keyediterator, see ExprFilter

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 5]
```diff
@@ -110,10 +206,13 @@ public static <A, B> int compare(A a, B b) throws IllegalArgumentException, Clas
...
 			return (Expression<? extends R>) this;
 
+		if (mappingExpr != null)
+			return super.getConvertedExpression(to);
+
 		Expression<? extends R> convertedList = list.getConvertedExpression(to);
 		if (convertedList != null)
 			//noinspection unchecked
-			return (Expression<? extends R>) new ExprSortedList(convertedList);
+			return (Expression<? extends R>) new ExprSortedList(convertedList, null, descendingOrder);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: why do we lose the mappingExpr here?

**Review by sovdeeth — CHANGES_REQUESTED** [review]
looks pretty good so far!

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 6]
```diff
@@ -152,14 +251,38 @@ public boolean isLoopOf(String input) {
...
+
+	@Override
+	public @UnknownNullability String getCurrentIndex() {
+		return currentIndex;
+	}
+
 	@Override
 	public String toString(@Nullable Event event, boolean debug) {
-		return "sorted " + list.toString(event, debug);
+		if (mappingExpr == null && !descendingOrder)
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java:298]: could definitely use a SyntaxStringBuilder

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java (hunk 5 above)
Comment (ethanrjs) [reply src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: that path is only reachable when `mappingExpr` is already null, covered by guard case prior. changed `null` to be `mappingExpr` for readability, thanks for pointing out!

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 7]
```diff
@@ -110,10 +220,14 @@ public static <A, B> int compare(A a, B b) throws IllegalArgumentException, Clas
 			//noinspection unchecked
 			return (Expression<? extends R>) this;
 
+		// when a mapping expression is present, InputSource wiring prevents safe shallow-copying
+		if (mappingExpr != null)
+			return super.getConvertedExpression(to);
+
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java:227]: yeah this probably isn't ideal but it seems a reasonable work-around.
I'd like to see some tests that certify it, though, like 
`if zombies within ((all entities in world "world") sorted by (input's name)) are x/y/z` where the sorted expr is forced to be converted in some manner

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk [hunk 8]
```diff
@@ -63,3 +63,25 @@ test "sorted expression with sort by":
 	set {_list::aaaaa} to 5
 	set {_sorted::*} to {_list::*} sorted by (length of input index)
 	assert {_sorted::*} is integers from 1 to 5 with "sort by index length failed"
+
+test "sorted by with type conversion":
+	# join requires %strings%, forcing getConvertedExpression(String.class) on the sorted-by expression.
+	# verifies InputSource is preserved through the ConvertedExpression wrapper.
+	set {_strs::*} to "fig", "apple", "banana" and "kiwi"
+	set {_joined} to join ({_strs::*} sorted by (length of input)) with ", "
+	assert {_joined} is "fig, kiwi, apple, banana" with "sorted-by type conversion failed (ascending)"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk]: unfortunately this doesn't actually convert anything (it's already strings)

File: src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk [hunk 9]
```diff
@@ -63,3 +63,25 @@ test "sorted expression with sort by":
...
+
+	# 'health of' requires %living entities%, forcing getConvertedExpression(LivingEntity.class)
+	spawn a zombie at test-location:
+		set {_z} to entity
+	spawn a pig at test-location:
+		set {_p} to entity
+	set {_entities::*} to {_z} and {_p}
+	set {_healths::*} to health of ({_entities::*} sorted by (health of input))
+	assert size of {_healths::*} is 2 with "sorted-by entity type conversion: wrong count"
+	assert first element of {_healths::*} <= last element of {_healths::*} with "sorted-by entity type conversion: not ascending"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk]: this one's alright, though I'd like one that doesn't use variables since variables have their own special handling for convertedExpressions

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk (hunk 8 above)
Comment (ethanrjs) [reply src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk]: oh my bad, will address this when i can. good catch

File: src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk (hunk 9 above)
Comment (ethanrjs) [reply src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk]: will do!

File: src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk [hunk 10]
```diff
@@ -64,24 +77,192 @@ test "sorted expression with sort by":
...
+			assert loop-index is "bob" with "keyed desc sort by: third index wrong"
+
+# --- keyed sort by with null mapped value ---
+
+test "sorted keyed with null mapping":
+	set {_data::a} to 1
+	set {_data::b} to 2
+	set {_data::c} to 3
+	catch runtime errors:
+		assert ({_data::*} sorted by ("%input%" parsed as time)) is not set with "keyed null mapping returned a value"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk:194]: should also assert the runtime error is what you expect it to be

File: src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk [hunk 11]
```diff
@@ -64,24 +77,192 @@ test "sorted expression with sort by":
...
+		set {_z2} to entity
+		set entity's name to "alpha"
+	spawn a zombie at test-location:
+		set {_z3} to entity
+		set entity's name to "bravo"
+
+	set {_names::*} to names of ((all entities in radius 2 around test-location) sorted by (input's name))
+	assert {_names::*} is "alpha", "bravo" and "charlie" with "sorted-by input name failed"
+	set {_healths::*} to health of ((all entities in radius 2 around test-location) sorted by (input's name))
+	assert size of {_healths::*} is 3 with "sorted-by entity conversion failed"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprSortedList.sk:242]: this is still a good test, but I would also add one that does `inventories within ()`
Bonus points for having a non-inventory holding entity like a minecart, and asserting that it's removed from the end list.

**Review by sovdeeth — CHANGES_REQUESTED** [review]
almost there!
